### PR TITLE
Feat/stackerdb sync

### PR DESCRIPTION
**Pull request by jcnelson**
This PR adds the StackerDB replication state machine, and connects it to the `PeerNetwork` and `Relayer` threads.  For now, it only adds basic unit tests, since this PR is already trending large.

### CODE REVIEW

**Review by kantai — COMMENTED** [review]
This LGTM, just had some mostly superficial comments.

File: stackslib/src/net/stackerdb/mod.rs [hunk 1]
```diff
@@ -198,3 +220,287 @@ pub struct SlotMetadata {
...
+            num_outbound_replicas,
+        })
+    }
+
+    /// Validate chunk data -- either pushed to us, or downloaded.
+    /// NOTE: does not check write frequency, since the caller has different ways of doing this.
+    /// Returns Ok(true) if the chunk is valid
+    /// Returns Ok(false) if the chunk is invalid
+    /// Returns Err(..) on DB error
+    pub fn validate_received_chunk(
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply stackslib/src/net/stackerdb/mod.rs:364]: Not without any savings -- the data being considered is loaded from different places, so there would need to be some extra boilerplate at both call sites to load it.

File: stackslib/src/net/stackerdb/sync.rs [hunk 2]
```diff
@@ -246,6 +249,8 @@ impl<NC: NeighborComms> StackerDBSync<NC> {
             .collect();
 
         schedule.sort_by(|item_1, item_2| item_1.1.len().cmp(&item_2.1.len()));
+        schedule.reverse();
+
```
Comment (kantai) [inline stackslib/src/net/stackerdb/sync.rs]: I think this instance was in the correct order before: when making requests _for_ data, you want to request in rarest-first order, which means ordering by which chunks have the least number of servers (`schedule` is a list of who _has_ the data). In `make_chunk_push_schedule`, `schedule` is a list of who _needs_ the data, so rarest-first order means ordering by the longest lists (where a `reverse()` call makes the order correct)

**Review by kantai — COMMENTED** [review]
LGTM -- just had one question about the request schedule ordering.

**Review by kantai — APPROVED** [review]

**Comment by jcnelson** [comment]
Ping @obycode ;)

**Review by obycode — APPROVED** [review]
LGTM! The documentation in here is appreciated 🙏

**Pull request merged**